### PR TITLE
Feat(Plugins): Download plugins over HTTP

### PR DESCRIPTION
**Pull request by link108**

### CODE REVIEW

**Comment by link108** [comment]
boo @armory-jenkins

File: kork-plugins/src/main/java/com/netflix/spinnaker/kork/plugins/spring/PluginDownloader.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    for (String pluginLocation : pluginLocations) {
+      urls.add(getUrlFromJarLocation(pluginLocation));
+    }
+
+    return urls.toArray(new URL[0]);
+  }
+
+  private URL getUrlFromJarLocation(String jarLocation) throws IOException {
+    if (jarLocation.matches("https?.*")) {
+      jarLocation = downloadJar("/Users/cmotevasselani/working/random/test/", jarLocation);
```
Comment (bpowell) [inline kork-plugins/src/main/java/com/netflix/spinnaker/kork/plugins/spring/PluginDownloader.java]: this has your path in it `/Users/cmotevasselani`

File: kork-plugins/src/main/java/com/netflix/spinnaker/kork/plugins/spring/PluginDownloader.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    Request request =
+        new com.squareup.okhttp.Request.Builder().url(jarLocation).method("GET", null).build();
+
+    Response response = client.newCall(request).execute();
+
+    InputStream is = response.body().byteStream();
+
+    String localFileName = pluginDirectory + FilenameUtils.getBaseName(jarLocation) + ".jar";
+    File localFile = new File(localFileName);
+    // localFile.createNewFile();
```
Comment (bpowell) [inline kork-plugins/src/main/java/com/netflix/spinnaker/kork/plugins/spring/PluginDownloader.java]: is this needed or should be removed?

**Review by bpowell — CHANGES_REQUESTED** [review]

File: kork-plugins/src/main/java/com/netflix/spinnaker/kork/plugins/spring/PluginDownloader.java (hunk 2 above)
Comment (link108) [reply kork-plugins/src/main/java/com/netflix/spinnaker/kork/plugins/spring/PluginDownloader.java]: can be removed

File: kork-plugins/src/main/java/com/netflix/spinnaker/kork/plugins/spring/PluginDownloader.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  /**
+   * This method returns the local filepath for a single jar. If the jar is not present, it will
+   * download the jar.
+   *
+   * @param jarLocation
+   * @return
+   * @throws IOException
+   */
+  public URL getUrlFromJarLocation(String jarLocation) throws IOException {
+    if (jarLocation.matches("https?.*")) {
```
Comment (ncknt) [inline kork-plugins/src/main/java/com/netflix/spinnaker/kork/plugins/spring/PluginDownloader.java:92]: Not sure I understand why you're donwloading the JARs by hand instead of passing the URLs to the `URLClassloader`. Is it not working? Because functionally (if it works) it's equivalent: if the JAR cannot load, the app won't start. It'd also obey any other options available (like security restrictions) - which may or may not be desirable.

If you really need to download first, I'd use parse with the URL and download or not based on the protocol. It'd at least validate the URL format early on and you could use `url.getPath()` etc. without needing `FilenameUtils` which seems to bring an extra dependency.

Comment (link108) [reply kork-plugins/src/main/java/com/netflix/spinnaker/kork/plugins/spring/PluginDownloader.java:92]: I tried previously, but let me try again... I have a better understanding of this now 👍 would be a better solution

**Comment by link108** [comment]
closing for https://github.com/armory-io/kork/pull/5
